// @urql/exchange-graphcache/dist/urql-exchange-graphcache.es.js
import { share, mergeMap, fromArray, take, buffer, fromPromise, empty, tap, map, concat, filter, merge } from "wonka";

import { stringifyVariables, createRequest, formatDocument } from "urql/core";

import { isWrappingType, Kind, valueFromASTUntyped, buildClientSchema, isNullableType, isNonNullType, isListType, GraphQLObjectType, GraphQLInterfaceType, GraphQLUnionType, visit, visitWithTypeInfo, TypeInfo, isCompositeType, isAbstractType } from "graphql";

function m() {
  return (m = Object.assign || function(a) {
    for (var b = 1; b < arguments.length; b++) {
      var d, c = arguments[b];
      for (d in c) {
        Object.prototype.hasOwnProperty.call(c, d) && (a[d] = c[d]);
      }
    }
    return a;
  }).apply(this, arguments);
}

function t(a) {
  return a.name.value;
}

function u(a) {
  return void 0 !== a.alias ? a.alias.value : t(a);
}

function v(a) {
  return void 0 !== a.selectionSet ? a.selectionSet.selections : [];
}

function aa(a) {
  return void 0 !== (a = a.typeCondition) ? t(a) : null;
}

function ba(a) {
  return isWrappingType(a) ? ba(a.ofType) : a || null;
}

var ca = new Set, A = [];

function B(a, b) {
  var c = "";
  b.kind === Kind.INLINE_FRAGMENT ? c = a ? 'Inline Fragment on "' + a + '"' : "Inline Fragment" : b.kind === Kind.OPERATION_DEFINITION ? c = (b.name ? '"' + b.name.value + '"' : "Unnamed") + " " + b.operation : b.kind === Kind.FRAGMENT_DEFINITION && (c = '"' + b.name.value + '" Fragment');
  c && A.push(c);
}

function da() {
  return A.length ? "\n(Caused At: " + A.join(", ") + ")" : "";
}

function C(a, b, c) {
  if (!a) {
    throw a = b || "Minfied Error #" + c + "\n", "production" !== process.env.NODE_ENV && (a += da()), 
    (c = Error(a + "\nhttps://github.com/FormidableLabs/urql-exchange-graphcache/blob/master/docs/help.md#" + c)).name = "Graphcache Error", 
    c;
  }
}

function D(a, b) {
  ca.has(a) || (console.warn(a + da() + "\nhttps://github.com/FormidableLabs/urql-exchange-graphcache/blob/master/docs/help.md#" + b), 
  ca.add(a));
}

function F(a, b) {
  return b ? a + "(" + stringifyVariables(b) + ")" : a;
}

function ea(a) {
  var b = a.indexOf("(");
  return -1 < b ? {
    fieldKey: a,
    fieldName: a.slice(0, b),
    arguments: JSON.parse(a.slice(b + 1, -1))
  } : {
    fieldKey: a,
    fieldName: a,
    arguments: null
  };
}

var fa = "production" === process.env.NODE_ENV && "undefined" != typeof Promise ? Promise.prototype.then.bind(Promise.resolve()) : function(a) {
  return setTimeout(a, 0);
};

function G() {
  return Object.create(null);
}

var H = null, I = null, J = null;

function ha() {
  return {
    optimistic: G(),
    base: new Map,
    keys: []
  };
}

function K(a, b) {
  H = a;
  I = new Set;
  J = b;
  "production" !== process.env.NODE_ENV && (A.length = 0);
}

function L() {
  var c = H;
  !c.gcScheduled && 0 < c.gcBatch.size && (c.gcScheduled = !0, fa((function a() {
    ia(c);
  })));
  c.storage && !c.persistenceScheduled && (c.persistenceScheduled = !0, fa((function b() {
    c.storage.write(c.persistenceBatch);
    c.persistenceScheduled = !1;
    c.persistenceBatch = G();
  })));
  J = I = H = null;
  "production" !== process.env.NODE_ENV && (A.length = 0);
}

function M() {
  C(null !== I, "production" !== process.env.NODE_ENV ? "Invalid Cache call: The cache may only be accessed or mutated duringoperations like write or query, or as part of its resolvers, updaters, or optimistic configs." : "", 2);
  return I;
}

function ka(a, b, c, d) {
  J ? (void 0 === a.optimistic[J] && (a.optimistic[J] = new Map, a.keys.unshift(J)), 
  a = a.optimistic[J]) : a = a.base;
  var e = a.get(b);
  void 0 === e && a.set(b, e = G());
  void 0 !== d || J ? e[c] = d : delete e[c];
}

function la(a, b, c) {
  for (var d = 0, e = a.keys.length; d < e; d++) {
    var f = a.optimistic[a.keys[d]].get(b);
    if (void 0 !== f && c in f) {
      return f[c];
    }
  }
  return void 0 !== (a = a.base.get(b)) ? a[c] : void 0;
}

function ma(a, b) {
  var c = a.keys.indexOf(b);
  -1 < c && (delete a.optimistic[b], a.keys.splice(c, 1));
}

function na(a, b, c, d) {
  var e = void 0 !== b[c] ? b[c] : 0;
  b = b[c] = e + d | 0;
  void 0 !== a && (0 >= b ? a.add(c) : 0 >= e && 0 < b && a.delete(c));
}

function oa(a, b, c, d) {
  if ("string" == typeof c) {
    na(a, b, c, d);
  } else if (Array.isArray(c)) {
    for (var e = 0, f = c.length; e < f; e++) {
      var g = c[e];
      g && na(a, b, g, d);
    }
  }
}

function pa(a, b, c) {
  if (void 0 !== c) {
    for (var d in c) {
      b.has(d) || (a.push(ea(d)), b.add(d));
    }
  }
}

function qa(a, b, c, d) {
  pa(a, b, d.base.get(c));
  for (var e = 0, f = d.keys.length; e < f; e++) {
    pa(a, b, d.optimistic[d.keys[e]].get(c));
  }
}

function ia(a) {
  a.gcScheduled = !1;
  a.gcBatch.forEach((function(b) {
    if (0 >= (a.refCount[b] || 0)) {
      for (var c in a.refLock) {
        var d = a.refLock[c];
        if (0 < (d[b] || 0)) {
          return;
        }
        delete d[b];
      }
      delete a.refCount[b];
      a.gcBatch.delete(b);
      if (void 0 !== (c = a.records.base.get(b)) && (a.records.base.delete(b), a.storage)) {
        for (var e in c) {
          a.persistenceBatch["r|" + b + "." + e] = void 0;
        }
      }
      if (void 0 !== (e = a.links.base.get(b))) {
        a.links.base.delete(b);
        for (var f in e) {
          a.storage && (a.persistenceBatch["l|" + b + "." + f] = void 0), oa(a.gcBatch, a.refCount, e[f], -1);
        }
      }
    } else {
      a.gcBatch.delete(b);
    }
  }));
}

function N(a, b) {
  "__typename" !== b && (a !== H.queryRootKey ? I.add(a) : void 0 !== b && I.add(a + "." + b));
}

function O(a, b) {
  N(a, b);
  return la(H.records, a, b);
}

function P(a, b) {
  N(a, b);
  return la(H.links, a, b);
}

function Q(a, b, c) {
  N(a, b);
  ka(H.records, a, b, c);
  H.storage && !J && (H.persistenceBatch["r|" + a + "." + b] = c);
}

function ra(a, b, c) {
  var d = H;
  if (J) {
    var e = d.refLock[J] || (d.refLock[J] = G());
    var f = d.links.optimistic[J];
  } else {
    d.storage && (d.persistenceBatch["l|" + a + "." + b] = c);
    e = d.refCount;
    f = d.links.base;
    var g = d.gcBatch;
  }
  f = void 0 !== (f = void 0 !== f ? f.get(a) : void 0) ? f[b] : null;
  N(a, b);
  ka(d.links, a, b, c);
  oa(g, e, f, -1);
  oa(g, e, c, 1);
}

function sa(a, b, c, d) {
  if (!b) {
    return !1;
  }
  var e = aa(a);
  if (b === e) {
    return !0;
  }
  "production" !== process.env.NODE_ENV && D("Heuristic Fragment Matching: A fragment is trying to match against the `" + b + "` type, but the type condition is `" + e + "`. Since GraphQL allows for interfaces `" + e + "` may be aninterface.\nA schema needs to be defined for this match to be deterministic, otherwise the fragment will be matched heuristically!", 16);
  return !v(a).some((function(a) {
    if (a.kind !== Kind.FIELD) {
      return !1;
    }
    a = F(t(a), R(a, d.variables));
    return !(void 0 !== O(c, a) || void 0 !== P(c, a));
  }));
}

function S(a, b, c, d) {
  this.typename = a;
  this.entityKey = b;
  this.context = d;
  this.indexStack = [ 0 ];
  this.selectionStack = [ c ];
}

S.prototype.next = function() {
  for (;0 !== this.indexStack.length; ) {
    var a = this.indexStack[this.indexStack.length - 1]++, b = this.selectionStack[this.selectionStack.length - 1];
    if (a >= b.length) {
      this.indexStack.pop(), this.selectionStack.pop();
    } else {
      a = b[a];
      a: {
        b = this.context.variables;
        var c = a.directives;
        if (void 0 !== c) {
          for (var d = 0, e = c.length; d < e; d++) {
            var f = c[d], g = t(f), h = "include" === g;
            if ((h || "skip" === g) && (f = f.arguments ? f.arguments[0] : null) && "if" === t(f) && ("boolean" == typeof (f = valueFromASTUntyped(f.value, b)) || null === f)) {
              b = h ? !!f : !f;
              break a;
            }
          }
        }
        b = !0;
      }
      if (b) {
        if (a.kind !== Kind.FIELD) {
          if (void 0 !== (a = a.kind !== Kind.INLINE_FRAGMENT ? this.context.fragments[t(a)] : a) && ("production" !== process.env.NODE_ENV && B(this.typename, a), 
          void 0 !== this.context.schemaPredicates ? this.context.schemaPredicates.isInterfaceOfType(aa(a), this.typename) : sa(a, this.typename, this.entityKey, this.context))) {
            this.indexStack.push(0), this.selectionStack.push(v(a));
          }
        } else if ("__typename" !== t(a)) {
          return a;
        }
      }
    }
  }
};

function T(a) {
  return void 0 === a ? null : a;
}

function ta(a, b, c) {
  K(a.data, 0);
  a = ua(a, b, c);
  L();
  return a;
}

function ua(a, b, c) {
  var d = U(b.query), e = {
    dependencies: M()
  }, f = v(d), g = a.getRootKey(d.operation);
  a = {
    parentTypeName: g,
    parentKey: g,
    parentFieldKey: "",
    fieldName: "",
    variables: V(d, b.variables),
    fragments: W(b.query),
    result: e,
    store: a,
    schemaPredicates: a.schemaPredicates
  };
  "production" !== process.env.NODE_ENV && B(g, d);
  g === a.store.getRootKey("query") ? va(a, g, f, c) : wa(a, g, f, c);
  return e;
}

function xa(a, b, c) {
  K(a.data, c);
  var d = U(b.query);
  c = {
    dependencies: M()
  };
  var e = a.getRootKey("mutation"), f = a.getRootKey(d.operation);
  C(f === e, "production" !== process.env.NODE_ENV ? "writeOptimistic(...) was called with an operation that is not a mutation.\nThis case is unsupported and should never occur." : "", 10);
  "production" !== process.env.NODE_ENV && B(f, d);
  a = {
    parentTypeName: e,
    parentKey: e,
    parentFieldKey: "",
    fieldName: "",
    variables: V(d, b.variables),
    fragments: W(b.query),
    result: c,
    store: a,
    schemaPredicates: a.schemaPredicates,
    optimistic: !0
  };
  b = G();
  d = new S(f, f, v(d), a);
  for (var g; void 0 !== (g = d.next()); ) {
    if (void 0 !== g.selectionSet) {
      var h = t(g), k = a.store.optimisticMutations[h];
      if (void 0 !== k) {
        a.fieldName = h;
        ya(a, T(k = k((f = R(g, a.variables)) || G(), a.store, a)), v(g));
        b[h] = k;
        void 0 !== (g = a.store.updates[e][h]) && g(b, f || G(), a.store, a);
      }
    }
  }
  L();
  return c;
}

function za(a, b, c, d) {
  b = W(b);
  var e = Object.keys(b);
  if (void 0 === (e = b[e[0]])) {
    return "production" !== process.env.NODE_ENV ? D("writeFragment(...) was called with an empty fragment.\nYou have to call it with at least one fragment in your GraphQL document.", 11) : void 0;
  }
  var f = e.typeCondition.name.value;
  c = m({
    __typename: f
  }, c);
  var g = a.keyOfEntity(c);
  if (!g) {
    return "production" !== process.env.NODE_ENV ? D("Can't generate a key for writeFragment(...) data.\nYou have to pass an `id` or `_id` field or create a custom `keys` config for `" + f + "`.", 12) : void 0;
  }
  "production" !== process.env.NODE_ENV && B(f, e);
  va(a = {
    parentTypeName: f,
    parentKey: g,
    parentFieldKey: "",
    fieldName: "",
    variables: d || {},
    fragments: b,
    result: {
      dependencies: M()
    },
    store: a,
    schemaPredicates: a.schemaPredicates
  }, g, v(e), c);
}

function va(a, b, c, d) {
  var e = b === a.store.getRootKey("query") ? b : d.__typename;
  if ("string" == typeof e) {
    Q(b, "__typename", e);
    c = new S(e, b, c, a);
    for (var f; void 0 !== (f = c.next()); ) {
      var g = t(f), h = R(f, a.variables);
      h = F(g, h);
      var k = d[u(f)], l = b + "." + h;
      if ("production" !== process.env.NODE_ENV) {
        if (void 0 === k) {
          g = a.optimistic ? "\nYour optimistic result may be missing a field!" : "";
          f = void 0 === f.selectionSet ? "scalar (number, boolean, etc)" : "selection set";
          "production" !== process.env.NODE_ENV && D("Invalid undefined: The field at `" + h + "` is `undefined`, but the GraphQL query expects a " + f + " for this field." + g, 13);
          continue;
        } else {
          a.schemaPredicates && e && a.schemaPredicates.isFieldAvailableOnType(e, g);
        }
      }
      void 0 === f.selectionSet ? Q(b, h, k) : (g = T(k), ra(b, h, f = Aa(a, l, v(f), g)));
    }
  }
}

function Aa(a, b, c, d) {
  if (Array.isArray(d)) {
    for (var e = Array(d.length), f = 0, g = d.length; f < g; f++) {
      var h = Aa(a, b + "." + f, c, d[f]);
      e[f] = h;
    }
    return e;
  }
  if (null === d) {
    return null;
  }
  f = null !== (e = a.store.keyOfEntity(d)) ? e : b;
  g = d.__typename;
  void 0 !== a.store.keys[d.__typename] || null !== e || "string" != typeof g || g.endsWith("Connection") || g.endsWith("Edge") || "PageInfo" === g || "production" !== process.env.NODE_ENV && D("Invalid key: The GraphQL query at the field at `" + b + "` has a selection set, but no key could be generated for the data at this field.\nYou have to request `id` or `_id` fields for all selection sets or create a custom `keys` config for `" + g + "`.\nEntities without keys will be embedded directly on the parent entity. If this is intentional, create a `keys` config for `" + g + "` that always returns null.", 15);
  va(a, f, c, d);
  return f;
}

function wa(a, b, c, d) {
  var e = b === a.store.getRootKey("mutation") || b === a.store.getRootKey("subscription");
  c = new S(b, b, c, a);
  for (var f; void 0 !== (f = c.next()); ) {
    var g = t(f), h = R(f, a.variables);
    var k = F(g, h);
    k = b + "." + k;
    if (void 0 !== f.selectionSet) {
      ya(a, T(d[u(f)]), v(f));
    }
    e && (a.parentTypeName = b, a.parentKey = b, a.parentFieldKey = k, a.fieldName = g, 
    void 0 !== (f = a.store.updates[b][g]) && f(d, h || G(), a.store, a));
  }
}

function ya(a, b, c) {
  if (Array.isArray(b)) {
    for (var d = Array(b.length), e = 0, f = b.length; e < f; e++) {
      d[e] = ya(a, b[e], c);
    }
    return d;
  }
  null !== b && (null !== (d = a.store.keyOfEntity(b)) ? va(a, d, c, b) : wa(a, b.__typename, c, b));
}

function X(a, b, c, d, e) {
  var g, f = this;
  this.gcScheduled = !1;
  this.gc = function() {
    ia(f.data);
    f.gcScheduled = !1;
  };
  this.keyOfField = F;
  this.resolvers = b || {};
  this.optimisticMutations = d || {};
  this.keys = e || {};
  this.schemaPredicates = a;
  this.updates = {
    Mutation: c && c.Mutation || {},
    Subscription: c && c.Subscription || {}
  };
  a ? (c = (b = a.schema).getQueryType(), a = b.getMutationType(), b = b.getSubscriptionType(), 
  this.rootFields = {
    query: c = c ? c.name : "Query",
    mutation: a = a ? a.name : "Mutation",
    subscription: b = b ? b.name : "Subscription"
  }, this.rootNames = ((g = {})[c] = "query", g[a] = "mutation", g[b] = "subscription", 
  g)) : (this.rootFields = {
    query: "Query",
    mutation: "Mutation",
    subscription: "Subscription"
  }, this.rootNames = {
    Query: "query",
    Mutation: "mutation",
    Subscription: "subscription"
  });
  this.data = function ja(a) {
    return {
      persistenceScheduled: !1,
      persistenceBatch: G(),
      gcScheduled: !1,
      queryRootKey: a,
      gcBatch: new Set,
      refCount: G(),
      refLock: G(),
      links: ha(),
      records: ha(),
      storage: null
    };
  }(this.getRootKey("query"));
}

X.prototype.getRootKey = function(a) {
  return this.rootFields[a];
};

X.prototype.keyOfEntity = function(a) {
  var b = a.__typename, c = a.id, d = a._id;
  if (!b) {
    return null;
  }
  if (void 0 !== this.rootNames[b]) {
    return b;
  }
  var e;
  this.keys[b] ? e = this.keys[b](a) : null != c ? e = "" + c : null != d && (e = "" + d);
  return e ? b + ":" + e : null;
};

X.prototype.resolveFieldByKey = function(a, b) {
  if (null === (a = null !== a && "string" != typeof a ? this.keyOfEntity(a) : a)) {
    return null;
  }
  var c = O(a, b);
  return void 0 !== c ? c : (b = P(a, b)) ? b : null;
};

X.prototype.resolve = function(a, b, c) {
  return this.resolveFieldByKey(a, F(b, c));
};

X.prototype.invalidateQuery = function(a, b) {
  !function Ba(a, b, c) {
    if ("Query" !== b) {
      var d = O(b, "__typename");
      if ("string" != typeof d) {
        return;
      }
      Q(b, "__typename", void 0);
    } else {
      d = b;
    }
    c = new S(d, b, c, a);
    for (var e; void 0 !== (e = c.next()); ) {
      var f = t(e), g = F(f, R(e, a.variables));
      "production" !== process.env.NODE_ENV && a.schemaPredicates && d && a.schemaPredicates.isFieldAvailableOnType(d, f);
      if (void 0 === e.selectionSet) {
        Q(b, g, void 0);
      } else if (e = v(e), f = P(b, g), ra(b, g, void 0), Q(b, g, void 0), Array.isArray(f)) {
        g = 0;
        for (var h = f.length; g < h; g++) {
          var k = f[g];
          null !== k && Ba(a, k, e);
        }
      } else {
        f && Ba(a, f, e);
      }
    }
  }(b = {
    variables: V(a = U((b = createRequest(a, b)).query), b.variables),
    fragments: W(b.query),
    store: this,
    schemaPredicates: this.schemaPredicates
  }, b.store.getRootKey("query"), v(a));
};

X.prototype.inspectFields = function(a) {
  if (null !== (a = null !== a && "string" != typeof a ? this.keyOfEntity(a) : a)) {
    var b = H.links, c = H.records, d = [], e = new Set;
    N(a);
    qa(d, e, a, b);
    qa(d, e, a, c);
    a = d;
  } else {
    a = [];
  }
  return a;
};

X.prototype.updateQuery = function(a, b) {
  a = createRequest(a.query, a.variables);
  null !== (b = b(this.readQuery(a))) && ua(this, a, b);
};

X.prototype.readQuery = function(a) {
  return Ca(this, createRequest(a.query, a.variables)).data;
};

X.prototype.readFragment = function(a, b, c) {
  a = W(a);
  var d = Object.keys(a);
  if (void 0 === (d = a[d[0]])) {
    "production" !== process.env.NODE_ENV && D("readFragment(...) was called with an empty fragment.\nYou have to call it with at least one fragment in your GraphQL document.", 6), 
    c = null;
  } else {
    var e = d.typeCondition.name.value;
    "string" == typeof b || b.__typename || (b.__typename = e);
    (b = "string" != typeof b ? this.keyOfEntity(m({
      __typename: e
    }, b)) : b) ? ("production" !== process.env.NODE_ENV && B(e, d), c = Y({
      parentTypeName: e,
      parentKey: b,
      parentFieldKey: "",
      fieldName: "",
      variables: c || {},
      fragments: a,
      partial: !1,
      store: this,
      schemaPredicates: this.schemaPredicates
    }, b, v(d), G()) || null) : ("production" !== process.env.NODE_ENV && D("Can't generate a key for readFragment(...).\nYou have to pass an `id` or `_id` field or create a custom `keys` config for `" + e + "`.", 7), 
    c = null);
  }
  return c;
};

X.prototype.writeFragment = function(a, b, c) {
  za(this, a, b, c);
};

function R(a, b) {
  if (void 0 === a.arguments || 0 === a.arguments.length) {
    return null;
  }
  for (var c = G(), d = 0, e = 0, f = a.arguments.length; e < f; e++) {
    var g = a.arguments[e], h = valueFromASTUntyped(g.value, b);
    null != h && (c[t(g)] = h, d++);
  }
  return 0 < d ? c : null;
}

function V(a, b) {
  if (void 0 === a.variableDefinitions) {
    return {};
  }
  var c = b || {};
  return a.variableDefinitions.reduce((function(a, b) {
    var d = t(b.variable), e = c[d];
    if (void 0 === e) {
      if (void 0 !== b.defaultValue) {
        e = valueFromASTUntyped(b.defaultValue, c);
      } else {
        return a;
      }
    }
    a[d] = e;
    return a;
  }), G());
}

function Z(a) {
  this.schema = buildClientSchema(a);
}

Z.prototype.isFieldNullable = function(a, b) {
  return void 0 === (a = Da(this.schema, a, b)) ? !1 : isNullableType(a.type);
};

Z.prototype.isListNullable = function(a, b) {
  if (void 0 === (a = Da(this.schema, a, b))) {
    return !1;
  }
  a = isNonNullType(a.type) ? a.type.ofType : a.type;
  return isListType(a) && isNullableType(a.ofType);
};

Z.prototype.isFieldAvailableOnType = function(a, b) {
  return !!Da(this.schema, a, b);
};

Z.prototype.isInterfaceOfType = function(a, b) {
  if (!b || !a) {
    return !1;
  }
  if (b === a) {
    return !0;
  }
  var c = this.schema.getType(a), d = this.schema.getType(b);
  if (c instanceof GraphQLObjectType) {
    return c === d;
  }
  C(c instanceof GraphQLInterfaceType || c instanceof GraphQLUnionType, "production" !== process.env.NODE_ENV ? "Invalid Abstract type: The type `" + a + "` is not an Interface or Union type in the defined schema, but a fragment in the GraphQL document is using it as a type condition." : "", 5);
  Ea(d, b);
  return this.schema.isPossibleType(c, d);
};

function Da(a, b, c) {
  Ea(a = a.getType(b), b);
  if (void 0 === (a = a.getFields()[c])) {
    "production" !== process.env.NODE_ENV && D("Invalid field: The field `" + c + "` does not exist on `" + b + "`, but the GraphQL document expects it to exist.\nTraversal will continue, however this may lead to undefined behavior!", 4);
  } else {
    return a;
  }
}

function Ea(a, b) {
  C(a instanceof GraphQLObjectType, "production" !== process.env.NODE_ENV ? "Invalid Object type: The type `" + b + "` is not an object in the defined schema, but the GraphQL document is traversing it." : "", 3);
}

function Fa(a) {
  return a.kind === Kind.FRAGMENT_DEFINITION;
}

function Ga(a) {
  return a.kind === Kind.OPERATION_DEFINITION;
}

function U(a) {
  C(!!(a = a.definitions.find(Ga)), "production" !== process.env.NODE_ENV ? "Invalid GraphQL document: All GraphQL documents must contain an OperationDefinitionnode for a query, subscription, or mutation." : "", 1);
  return a;
}

function Ha(a, b) {
  a[t(b)] = b;
  return a;
}

function W(a) {
  return a.definitions.filter(Fa).reduce(Ha, {});
}

function Ia(a, b, c) {
  K(a.data, 0);
  a = Ca(a, b, c);
  L();
  return a;
}

function Ca(a, b, c) {
  var d = U(b.query), e = a.getRootKey(d.operation), f = v(d);
  a = {
    parentTypeName: e,
    parentKey: e,
    parentFieldKey: "",
    fieldName: "",
    variables: V(d, b.variables),
    fragments: W(b.query),
    partial: !1,
    store: a,
    schemaPredicates: a.schemaPredicates
  };
  "production" !== process.env.NODE_ENV && B(e, d);
  c = c || G();
  c = e !== a.store.getRootKey("query") ? Ja(a, e, f, c) : Y(a, e, f, c);
  return {
    dependencies: M(),
    partial: void 0 === c ? !1 : a.partial,
    data: void 0 === c ? null : c
  };
}

function Ja(a, b, c, d) {
  if ("string" != typeof d.__typename) {
    return d;
  }
  b = new S(b, b, c, a);
  (c = G()).__typename = d.__typename;
  for (var e; void 0 !== (e = b.next()); ) {
    var f = u(e), g = d[f];
    void 0 !== e.selectionSet && null !== g ? (g = T(g), c[f] = Ka(a, v(e), g)) : c[f] = g;
  }
  return c;
}

function Ka(a, b, c) {
  if (Array.isArray(c)) {
    for (var d = Array(c.length), e = 0, f = c.length; e < f; e++) {
      d[e] = Ka(a, b, c[e]);
    }
    return d;
  }
  if (null === c) {
    return null;
  }
  return null !== (d = a.store.keyOfEntity(c)) ? void 0 === (a = Y(a, d, b, G())) ? null : a : Ja(a, c.__typename, b, c);
}

function Y(a, b, c, d) {
  var e = a.store, f = a.schemaPredicates, g = b === e.getRootKey("query"), h = g ? b : O(b, "__typename");
  if ("string" == typeof h) {
    d.__typename = h;
    c = new S(h, b, c, a);
    for (var k, l = !1, p = !1; void 0 !== (k = c.next()); ) {
      var r = t(k), q = R(k, a.variables), z = u(k), E = F(r, q), w = O(b, E), n = b + "." + E;
      "production" !== process.env.NODE_ENV && f && h && f.isFieldAvailableOnType(h, r);
      var x = void 0, y = e.resolvers[h];
      if (void 0 !== y && "function" == typeof y[r]) {
        if (a.parentTypeName = h, a.parentKey = b, a.parentFieldKey = n, a.fieldName = r, 
        void 0 !== w && (d[z] = w), x = y[r](d, q || G(), e, a), void 0 !== k.selectionSet && (x = La(a, h, r, n, v(k), d[z] || G(), x)), 
        void 0 !== f && null === x && !f.isFieldNullable(h, r)) {
          return;
        }
      } else {
        void 0 === k.selectionSet ? x = w : void 0 !== (q = P(b, E)) ? x = Ma(a, q, h, r, v(k), d[z]) : "object" == typeof w && null !== w && (x = w);
      }
      if (void 0 === x && void 0 !== f && f.isFieldNullable(h, r)) {
        p = !0, d[z] = null;
      } else {
        if (void 0 === x) {
          return;
        }
        l = !0;
        d[z] = x;
      }
    }
    p && (a.partial = !0);
    return g && p && !l ? void 0 : d;
  }
}

function La(a, b, c, d, e, f, g) {
  if (Array.isArray(g)) {
    var h = a.schemaPredicates;
    h = void 0 === h || h.isListNullable(b, c);
    for (var k = Array(g.length), l = 0, p = g.length; l < p; l++) {
      var r = La(a, b, c, d + "." + l, e, void 0 !== f ? f[l] : void 0, g[l]);
      if (void 0 !== r || h) {
        k[l] = void 0 !== r ? r : null;
      } else {
        return;
      }
    }
    return k;
  }
  if (null == g) {
    return g;
  }
  if ("string" == typeof g || "object" == typeof g && "string" == typeof g.__typename) {
    b = void 0 === f ? G() : f;
    if ("string" == typeof g) {
      a = Y(a, g, e, b);
    } else {
      a: if (c = a.schemaPredicates, d = a.store.keyOfEntity(g) || d, h = g.__typename, 
      f = O(d, "__typename") || h, "string" != typeof f || h && f !== h) {
        "production" !== process.env.NODE_ENV && D("Invalid resolver data: The resolver at `" + d + "` returned an invalid typename that could not be reconciled with the cache.", 8), 
        a = void 0;
      } else {
        b.__typename = f;
        e = new S(f, d, e, a);
        for (l = k = !1; void 0 !== (h = e.next()); ) {
          p = t(h);
          r = u(h);
          var q = F(p, R(h, a.variables)), z = d + "." + q, E = O(d, q), w = g[p];
          "production" !== process.env.NODE_ENV && c && f && c.isFieldAvailableOnType(f, p);
          var n = void 0;
          void 0 !== w && void 0 === h.selectionSet ? n = w : void 0 === h.selectionSet ? n = E : void 0 !== w ? n = La(a, f, p, z, v(h), b[r], w) : void 0 !== (q = P(d, q)) ? n = Ma(a, q, f, p, v(h), b[r]) : "object" == typeof E && null !== E && (n = E);
          if (void 0 === n && void 0 !== c && c.isFieldNullable(f, p)) {
            l = !0, b[r] = null;
          } else if (void 0 === n) {
            a = void 0;
            break a;
          } else {
            k = !0, b[r] = n;
          }
        }
        l && (a.partial = !0);
        a = k ? b : void 0;
      }
    }
    return a;
  }
  "production" !== process.env.NODE_ENV && D("Invalid resolver value: The field at `" + d + "` is a scalar (number, boolean, etc), but the GraphQL query expects a selection set for this field.", 9);
}

function Ma(a, b, c, d, e, f) {
  if (Array.isArray(b)) {
    var g = a.schemaPredicates;
    g = void 0 !== g && g.isListNullable(c, d);
    for (var h = Array(b.length), k = 0, l = b.length; k < l; k++) {
      var p = Ma(a, b[k], c, d, e, void 0 !== f ? f[k] : void 0);
      if (void 0 !== p || g) {
        h[k] = void 0 !== p ? p : null;
      } else {
        return;
      }
    }
    return h;
  }
  return null === b ? null : Y(a, b, e, void 0 === f ? G() : f);
}

function Na(a, b) {
  return m(m({}, a), {
    context: m(m({}, a.context), {
      meta: m(m({}, a.context.meta), {
        cacheOutcome: b
      })
    })
  });
}

function Oa(a) {
  return m(m({}, a), {
    query: formatDocument(a.query)
  });
}

function Pa(a) {
  return "query" === a.operationName && "network-only" !== a.context.requestPolicy;
}

function Qa(a, b) {
  return m(m({}, a), {
    context: m(m({}, a.context), {
      requestPolicy: b
    })
  });
}

function Ra(a) {
  return Pa(a);
}

function Sa(a) {
  return Na(a.operation, a.outcome);
}

function Ta(a) {
  return "miss" === a.outcome;
}

function Ua(a) {
  return "miss" !== a.outcome;
}

function Va(a) {
  return !Pa(a);
}

function Ya(a) {
  return "populate" !== t(a);
}

function Za(a, b) {
  "FragmentDefinition" === b.kind && a.add(b.name.value);
  return a;
}

function $a(a, b, c, d) {
  function e(a, b) {
    if (!(b = c[b.name])) {
      return a;
    }
    for (var e = 0, f = b.length; e < f; e++) {
      for (var l = b[e].fragment, p = t(l), w = ab(l), n = 0, x = w.length; n < x; n++) {
        var y = w[n];
        k.has(y) || (g[y] = d[y]);
      }
      h[p] = l;
      a.push({
        kind: Kind.FRAGMENT_SPREAD,
        name: Xa(p)
      });
    }
    return a;
  }
  var f = new TypeInfo(a), g = G(), h = G(), k = new Set;
  return visit(b, visitWithTypeInfo(f, {
    Field: {
      enter: function(b) {
        if (b.directives) {
          var c = b.directives.filter(Ya);
          if (c.length !== b.directives.length) {
            var d = ba(f.getType());
            isCompositeType(d) ? d = isAbstractType(d) ? a.getPossibleTypes(d) : [ d ] : ("production" !== process.env.NODE_ENV && D("Invalid type: The type ` + type + ` is used with @populate but does not exist.", 17), 
            d = []);
            d = d.reduce(e, []);
            var g = v(b);
            d = 0 !== g.length + d.length ? d.concat(g) : [ {
              kind: Kind.FIELD,
              name: Xa("__typename")
            } ];
            return m(m({}, b), {
              directives: c,
              selectionSet: {
                kind: Kind.SELECTION_SET,
                selections: d
              }
            });
          }
        }
      }
    },
    Document: {
      enter: function(a) {
        a.definitions.reduce(Za, k);
      },
      leave: function(a) {
        var c, b = [].concat(a.definitions);
        for (c in h) {
          b.push(h[c]);
        }
        for (var d in g) {
          b.push(g[d]);
        }
        return m(m({}, a), {
          definitions: b
        });
      }
    }
  }));
}

function Xa(a) {
  return {
    kind: Kind.NAME,
    value: a
  };
}

function ab(a) {
  var b = [];
  visit(a, {
    FragmentSpread: function(a) {
      b.push(t(a));
    }
  });
  return b;
}

var Store = X;

var cacheExchange = function(a) {
  return function(b) {
    function c(a) {
      var b = a.operation, c = a.error, d = a.extensions, f = "query" === b.operationName, y = a.data, k = b.key;
      if (w.has(k)) {
        w.delete(k);
        var z = q.data;
        delete z.refLock[k];
        ma(z.records, k);
        ma(z.links, k);
      }
      if (null != y) {
        var l = ta(q, b, y).dependencies;
        if (f) {
          var n = Ia(q, b);
          y = n.data;
          n = n.dependencies;
        } else {
          y = Ia(q, b, y).data;
        }
      }
      h(k = new Set, l);
      f && h(k, n);
      g(a.operation, k);
      f && void 0 !== n && e(a.operation, n);
      return {
        data: y,
        error: c,
        extensions: d,
        operation: b
      };
    }
    function d(a) {
      var b = Ia(q, a), c = b.data, d = b.dependencies;
      b = b.partial;
      null === c ? d = "miss" : (e(a, d), d = b && "cache-only" !== a.context.requestPolicy ? "partial" : "hit");
      return {
        outcome: d,
        operation: a,
        data: c
      };
    }
    function e(a, b) {
      b.forEach((function(b) {
        (x[b] || (x[b] = [])).push(a.key);
        n.has(a.key) || n.set(a.key, "network-only" === a.context.requestPolicy ? Qa(a, "cache-and-network") : a);
      }));
    }
    function f(a) {
      if ("mutation" === a.operationName && "network-only" !== a.context.requestPolicy) {
        var b = a.key, c = xa(q, a, b).dependencies;
        0 !== c.size && (w.add(b), h(b = new Set, c), g(a, b));
      }
    }
    function g(a, b) {
      b.forEach((function(b) {
        if (b !== a.key) {
          var c = n.get(b);
          void 0 !== c && (n.delete(b), r.reexecuteOperation(Qa(c, "cache-first")));
        }
      }));
    }
    function h(a, b) {
      void 0 !== b && b.forEach((function c(b) {
        var c = x[b];
        if (void 0 !== c) {
          x[b] = [];
          b = 0;
          for (var d = c.length; b < d; b++) {
            a.add(c[b]);
          }
        }
      }));
    }
    function l(a) {
      var b = a.operation, c = a.outcome, d = b.context.requestPolicy;
      a = {
        operation: Na(b, c),
        data: a.data,
        error: a.error,
        extensions: a.extensions
      };
      if ("cache-and-network" === d || "cache-first" === d && "partial" === c) {
        a.stale = !0, r.reexecuteOperation(Qa(b, "network-only"));
      }
      return a;
    }
    var p = b.forward, r = b.client;
    a || (a = {});
    var q = new X(a.schema ? new Z(a.schema) : void 0, a.resolvers, a.updates, a.optimistic, a.keys);
    if (a.storage) {
      var z = a.storage;
      var E = z.read().then((function k(a) {
        var b = q.data, c = z;
        K(b, 0);
        for (var d in a) {
          var e = d.indexOf("."), f = d.slice(2, e);
          e = d.slice(e + 1);
          switch (d.charCodeAt(0)) {
           case 108:
            ra(f, e, a[d]);
            break;

           case 114:
            Q(f, e, a[d]);
          }
        }
        L();
        b.storage = c;
      }));
    }
    var w = new Set, n = new Map, x = G();
    return function(a) {
      a = share(a);
      var b = E ? mergeMap(fromArray)(take(1)(buffer(fromPromise(E))(a))) : empty;
      a = share(tap(f)(map(Oa)(concat([ b, a ]))));
      var e = share(map(d)(filter(Ra)(a)));
      b = map(Sa)(filter(Ta)(e));
      e = map(l)(filter(Ua)(e));
      a = map(c)(p(merge([ filter(Va)(a), b ])));
      return merge([ a, e ]);
    };
  };
};

var clearDataState = L;

var initDataState = K;

var populateExchange = function(a) {
  var b = a.schema;
  return function(a) {
    function c(a) {
      "teardown" === a.operationName && p.delete(a.key);
    }
    function e(a) {
      var b = a.key, c = a.query;
      if ("query" === a.operationName && (p.add(b), !l.has(b))) {
        l.add(b);
        a = function Wa(a, b) {
          var c = [], d = [], e = new TypeInfo(a);
          visit(b, visitWithTypeInfo(e, {
            Field: function(a) {
              if (a.selectionSet) {
                var b = ba(e.getType());
                C(b && !isAbstractType(b), "production" !== process.env.NODE_ENV ? "Invalid TypeInfo state: Found no flat schema type when one was expected." : "", 18);
                b = b.toString();
                d.push({
                  kind: Kind.FRAGMENT_DEFINITION,
                  typeCondition: {
                    kind: Kind.NAMED_TYPE,
                    name: Xa(b)
                  },
                  name: Xa(b + "_PopulateFragment_"),
                  selectionSet: a.selectionSet
                });
              }
            },
            FragmentDefinition: function(a) {
              c.push(a);
            }
          }));
          return [ c, d ];
        }(k, c);
        c = a[0];
        a = a[1];
        for (var d = 0, e = c.length; d < e; d++) {
          var f = c[d];
          r[t(f)] = f;
        }
        c = 0;
        for (d = a.length; c < d; c++) {
          f = t((e = a[c]).typeCondition), f = q[f] || (q[f] = []), e.name.value += f.length, 
          f.push({
            key: b,
            fragment: e
          });
        }
      }
    }
    function f(a) {
      if ("mutation" !== a.operationName) {
        return a;
      }
      var c, b = G();
      for (c in q) {
        b[c] = q[c].filter(g);
      }
      return m(m({}, a), {
        query: $a(k, a.query, b, r)
      });
    }
    function g(a) {
      return p.has(a.key);
    }
    var h = a.forward, k = buildClientSchema(b), l = new Set, p = new Set, r = G(), q = G();
    return function(a) {
      return h(map(f)(tap(c)(tap(e)(a))));
    };
  };
};

var query = Ia;

var read = Ca;

var write = ta;

var writeFragment = za;

var writeOptimistic = xa;

export { Store, cacheExchange, clearDataState, initDataState, populateExchange, query, read, write, writeFragment, writeOptimistic };
