// @urql/exchange-graphcache/dist/urql-exchange-graphcache-extras.es.js
import { stringifyVariables } from "urql/core";

function r() {
  return (r = Object.assign || function(a) {
    for (var d = 1; d < arguments.length; d++) {
      var c, e = arguments[d];
      for (c in e) {
        Object.prototype.hasOwnProperty.call(e, c) && (a[c] = e[c]);
      }
    }
    return a;
  }).apply(this, arguments);
}

var u = {
  __typename: "PageInfo",
  endCursor: null,
  startCursor: null,
  hasNextPage: !1,
  hasPreviousPage: !1
};

function x(a) {
  return "string" == typeof a ? a : null;
}

function z(a, d, e) {
  for (var c = new Set, b = 0, g = d.length; b < g; b++) {
    var p = a.resolve(d[b], "node");
    "string" == typeof p && c.add(p);
  }
  d = d.slice();
  b = 0;
  for (g = e.length; b < g; b++) {
    var k = a.resolve(p = e[b], "node");
    "string" != typeof k || c.has(k) || (c.add(k), d.push(p));
  }
  return d;
}

function B(a, d) {
  for (var e in d) {
    if ("first" !== e && "last" !== e && "after" !== e && "before" !== e) {
      if (!(e in a)) {
        return !1;
      }
      var c = a[e], b = d[e];
      if (typeof c != typeof b || "object" != typeof c ? c !== b : stringifyVariables(c) !== stringifyVariables(b)) {
        return !1;
      }
    }
  }
  return !0;
}

function C(a, d, e) {
  var c = x(a.resolveFieldByKey(d, e));
  if (!c) {
    return null;
  }
  d = a.resolve(c, "__typename");
  e = a.resolve(c, "edges") || [];
  if ("string" != typeof d) {
    return null;
  }
  d = {
    __typename: d,
    edges: e,
    pageInfo: u
  };
  var b = a.resolve(c, "pageInfo");
  if ("string" == typeof b) {
    c = x(a.resolve(b, "__typename"));
    var g = x(a.resolve(b, "endCursor")), p = x(a.resolve(b, "startCursor")), k = a.resolve(b, "hasNextPage");
    b = a.resolve(b, "hasPreviousPage");
    null === (c = d.pageInfo = {
      __typename: "string" == typeof c ? c : "PageInfo",
      hasNextPage: "boolean" == typeof k ? k : !!g,
      hasPreviousPage: "boolean" == typeof b ? b : !!p,
      endCursor: g,
      startCursor: p
    }).endCursor && (g = e[e.length - 1]) && (g = a.resolve(g, "cursor"), c.endCursor = x(g));
    null === c.startCursor && (e = e[0]) && (a = a.resolve(e, "cursor"), c.startCursor = x(a));
  }
  return d;
}

var relayPagination = function(a) {
  void 0 === a && (a = {});
  var d = a.mergeMode || "inwards";
  return function(a, c, b, g) {
    var e = g.fieldName, k = b.inspectFields(a = g.parentKey).filter((function(b) {
      return b.fieldName === e;
    })), A = k.length;
    if (0 !== A) {
      for (var w = null, q = [], n = [], h = r({}, u), m = 0; m < A; m++) {
        var l = k[m], f = l.fieldKey;
        null !== (l = l.arguments) && B(c, l) && (null !== (f = C(b, a, f)) && ("inwards" === d && "number" == typeof l.last && "number" == typeof l.first ? (h = f.edges.slice(0, l.first + 1), 
        l = f.edges.slice(-l.last), q = z(b, q, h), n = z(b, l, n), h = f.pageInfo) : l.after ? (q = z(b, q, f.edges), 
        h.endCursor = f.pageInfo.endCursor, h.hasNextPage = f.pageInfo.hasNextPage) : l.before ? (n = z(b, f.edges, n), 
        h.startCursor = f.pageInfo.startCursor, h.hasPreviousPage = f.pageInfo.hasPreviousPage) : ("number" == typeof l.last ? n = z(b, n, f.edges) : q = z(b, q, f.edges), 
        h = f.pageInfo), f.pageInfo.__typename !== h.__typename && (h.__typename = f.pageInfo.__typename), 
        w !== f.__typename && (w = f.__typename)));
      }
      if ("string" == typeof w) {
        if (!x(b.resolve(a, e, c))) {
          if (void 0 === g.schemaPredicates) {
            return;
          }
          g.partial = !0;
        }
        return {
          __typename: w,
          edges: "inwards" === d ? z(b, q, n) : z(b, n, q),
          pageInfo: {
            __typename: h.__typename,
            endCursor: h.endCursor,
            startCursor: h.startCursor,
            hasNextPage: h.hasNextPage,
            hasPreviousPage: h.hasPreviousPage
          }
        };
      }
    }
  };
};

var simplePagination = function(a) {
  function d(b, a) {
    for (var d in a) {
      if (d !== e && d !== c) {
        if (!(d in b)) {
          return !1;
        }
        var k = b[d], g = a[d];
        if (typeof k != typeof g || "object" != typeof k ? k !== g : stringifyVariables(k) !== stringifyVariables(g)) {
          return !1;
        }
      }
    }
    return !0;
  }
  void 0 === a && (a = {});
  var e = a.offsetArgument;
  void 0 === e && (e = "skip");
  var c = a.limitArgument;
  void 0 === c && (c = "limit");
  return function(b, a, c, k) {
    var g = k.fieldName, p = c.inspectFields(b = k.parentKey).filter((function(a) {
      return a.fieldName === g;
    })), q = p.length;
    if (0 !== q) {
      for (var n = new Set, h = [], m = null, l = 0; l < q; l++) {
        var f = p[l], t = f.fieldKey;
        if (null !== (f = f.arguments) && d(a, f) && (t = c.resolveFieldByKey(b, t), f = f[e], 
        null !== t && 0 !== t.length && "number" == typeof f)) {
          if (!m || f > m) {
            for (m = 0; m < t.length; m++) {
              var v = t[m];
              n.has(v) || (h.push(v), n.add(v));
            }
          } else {
            m = [];
            for (v = 0; v < t.length; v++) {
              var y = t[v];
              n.has(y) || (m.push(y), n.add(y));
            }
            h = m.concat(h);
          }
          m = f;
        }
      }
      if (c.resolve(b, g, a)) {
        return h;
      }
      if (void 0 !== k.schemaPredicates) {
        return k.partial = !0, h;
      }
    }
  };
};

export { relayPagination, simplePagination };
